src/Workspaces/Remote/ServiceHub/Services/RemoteHostService.cs

```diff
@@ -305,5 +307,61 @@ public Task SynchronizeGlobalAssetsAsync(Checksum[] checksums, CancellationToken
                 }
             }, cancellationToken);
         }
+
+        public Task SynchronizeTextAsync(DocumentId documentId, Checksum baseTextChecksum, IEnumerable<TextChange> textChanges, CancellationToken cancellationToken)
+        {
+            return RunServiceAsync(async token =>
+            {
+                using (RoslynLogger.LogBlock(FunctionId.RemoteHostService_SynchronizeTextAsync, Checksum.GetChecksumLogInfo, baseTextChecksum, token))
+                {
+                    var service = SolutionService.PrimaryWorkspace.Services.GetService<ISerializerService>();
+                    if (service == null)
+                    {
+                        return;
+                    }
+
+                    var text = await GetSourceText().ConfigureAwait(false);
+                    if (text == null)
+                    {
+                        // it won't bring in base text if it is not there already.
+                        // text needed will be pulled in when there is request
+                        return;
+                    }
+
+                    var newText = text.WithChanges(textChanges);
+                    var newChecksum = service.CreateChecksum(newText, token);
+
+                    // save new text in the cache so that when asked, the data is most likely already there
+                    AssetStorage.TryAddAsset(newChecksum, newText);
+                }
+
+                async Task<SourceText> GetSourceText()
```

Comment:
GetSourceTextAsync

Reply:
or TryGetSourceTextASync as this can fail.

Reply:
sure